libraries/AP_Common/AP_ExpandingArray.h

```diff
@@ -0,0 +1,128 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * ExpandingArray class description
+ *
+ * Elements are organised into "chunks" with each chunk holding "chunk_size" elements
+ * The "chunk_ptrs" array holds pointers to all allocated chunks
+ *
+ * The "expand" function allows expanding the array by a specified number of chunks
+ * The "expand_to_hold" function expands the array (if necessary) to hold at least the specified number of elements
+ *
+ * When the array is expanded up to two memory allocations are required:
+ *    1. if the chunk_ptrs array (which holds points to all allocated chunks) is full, this array will be re-allocated.
+ *       During this operation a new copy of the chunk_ptr array will be created with "chunk_ptr_increment" more rows,
+ *       the old array's data will be copied to the new array and finally the old array will be freed.
+ *    2. a new chunk will be allocated and a pointer to this new chunk will be added to the chunk_ptrs array
+ *
+ * Warnings:
+ *    1. memset, memcpy, memcmp cannot be used because the individual elements are not guaranteed to be next to each other in memory
+ *    2. operator[] functions do not perform any range checking so max_items() should be used when necessary to avoid out-of-bound memory access
+ */
+
+#pragma once
+
+#include <AP_Common/AP_Common.h>
+
+template <typename T>
+class AP_ExpandingArray
+{
+public:
+
+    AP_ExpandingArray<T>(uint16_t elements_per_chunk) :
+        chunk_size(elements_per_chunk)
+    {}
+
+    /* Do not allow copies */
+    AP_ExpandingArray<T>(const AP_ExpandingArray<T> &other) = delete;
+    AP_ExpandingArray<T> &operator=(const AP_ExpandingArray<T>&) = delete;
+
+    // current maximum number of items (using expand may increase this)
+    uint16_t max_items() const { return chunk_size * chunk_count; }
+
+    // allow use as an array for assigning to elements. no bounds checking is performed
+    T &operator[](uint16_t i)
+    {
+        const uint16_t chunk_num = i / chunk_size;
+        const uint16_t chunk_index = i % chunk_size;
+        return chunk_ptrs[chunk_num][chunk_index];
+    }
+
+    // allow use as an array for accessing elements. no bounds checking is performed
+    const T &operator[](uint16_t i) const
+    {
+        const uint16_t chunk_num = i / chunk_size;
+        const uint16_t chunk_index = i % chunk_size;
+        return chunk_ptrs[chunk_num][chunk_index];
+    }
+
+    // expand the array by specified number of chunks, returns true on success
+    bool expand(uint16_t num_chunks = 1)
```

Comment:
const num_chunks

Reply:
I'm a little worried that this would send us down the path of putting "const" in front of all arguments.. it's really not necessary and I think it hurts readability so I think I might not make this change.